### DISCUSSION TITLE
How did you learn how to add variable renaming functionality?

### DISCUSSION THREAD

**Discussion by hirehamir** [Beaglefoot/awk-language-server#16] [Q&A] [answered]
I saw commit 1b556efc23144db3b59e0ca02d4d6afd9931d8a0 and some of the implementation for adding variable renaming functionality.

Where did you learn how to do that? Are there any guides you recommend?

[The Language Server Protocol's specification has a section called "Rename Request"](https://microsoft.github.io/language-server-protocol/specification#textDocument_rename) with some details on what the request should look like, but no examples of what that looks like in action, in terms of what code in `client.ts` and `server.ts` achieve that functionality.

**Comment by Beaglefoot** [comment, marked as answer]
I figured it out on my own. And I'm unaware of any guide that covers such functionality.

Specification just tells you what you need to create in terms of request/response and related interfaces. The implementation is up to you.

It's not that difficult, but might get lengthy depending on your language. You basically need to build a graph of dependencies or who imports who. Then you locate the correct scope within each of you dependent documents. And finally you generate a list of edits.

You need some knowledge of algorithms to do all these steps correctly. At least BFS and DFS graph traversal. Note that if your language supports cyclical dependencies it's up to you to resolve them. You also need to test you solution extensively because there are too many edge cases and variations. Having no tests most probably will end up with situation where fixing one thing breaks the other.

**Reply by hirehamir** [reply]
I see. Thank you for the detailed response!
